accumulator/accumulate_signals_workflow.go

```diff
@@ -0,0 +1,159 @@
+		return 0, nil
+	}
+	if SignalToSignalTimeout < totalLeft {
+		return SignalToSignalTimeout, nil
+	}
+	return totalLeft, nil
+}
+
+// AccumulateSignalsWorkflow workflow definition
+func AccumulateSignalsWorkflow(ctx workflow.Context, bucketKey string, greetingsCAN []AccumulateGreeting, uniqueGreetingKeysMapCAN map[string]bool) (allGreetings string, err error) {
+	log := workflow.GetLogger(ctx)
+	var a AccumulateGreeting
+	greetings := []AccumulateGreeting{}
+	greetings = append(greetings, greetingsCAN...)
+	unprocessedGreetings := []AccumulateGreeting{}
+	uniqueGreetingKeysMap := make(map[string]bool)
+	for k, v := range uniqueGreetingKeysMapCAN {
+		uniqueGreetingKeysMap[k] = v
+	}
+	var FirstSignalTime time.Time
+	ExitRequested := false
+
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	// Listen to signals in a different goroutine
+	workflow.Go(ctx, func(gCtx workflow.Context) {
+		for {
+			log.Debug("in workflow.go signals goroutine for{} \n")
+			selector := workflow.NewSelector(gCtx)
+			selector.AddReceive(workflow.GetSignalChannel(gCtx, "greeting"), func(c workflow.ReceiveChannel, more bool) {
+				c.Receive(gCtx, &a)
+				unprocessedGreetings = append(unprocessedGreetings, a)
+				log.Debug("Signal Received with text: " + a.GreetingText + ", more: " + strconv.FormatBool(more) + "\n")
+				
+				a = AccumulateGreeting{} 
+			})
+			selector.AddReceive(workflow.GetSignalChannel(gCtx, "exit"), func(c workflow.ReceiveChannel, more bool) {
+				c.Receive(gCtx, nil)
+				ExitRequested = true
+				log.Debug("Exit Signal Received\n")
+			})
+			log.Debug("before select, greeting  text: " + a.GreetingText + "\n")
+			selector.Select(gCtx)
+			log.Debug("after select, greeting  text: " + a.GreetingText + "\n")
+			if FirstSignalTime.IsZero() {
+				FirstSignalTime = workflow.Now(gCtx)
+			}
+		}
+	})
+
+	for ; !workflow.GetInfo(ctx).GetContinueAsNewSuggested() ;{
+		// Wait for Signal or timeout
+		timeout, err := a.GetNextTimeout(ctx, ExitRequested, FirstSignalTime)
+
+		if err != nil {
+			log.Warn("error awaiting signal\n")
+			return "", err
+		}
+		if timeout <= 0 {
+			// do final processing? or just check for signal one more time
+			log.Debug("No time left for signals, checking one more time\n")
+		}
+
+		log.Debug("Awaiting for " + timeout.String() + "\n")
+		gotSignalBeforeTimeout, _ := workflow.AwaitWithTimeout(ctx, timeout, func() bool {
+			return len(unprocessedGreetings) > 0 || ExitRequested
+		})
+
+		// timeout happened without a signal coming in, so let's process the greetings and wrap it up!
+		if len(unprocessedGreetings) == 0 {
+			log.Debug("Into final processing, signal received? " + strconv.FormatBool(gotSignalBeforeTimeout)  + ", exit requested: " + strconv.FormatBool(ExitRequested) +"\n")
+			// do final processing
+			allGreetings = ""
+			err := workflow.ExecuteActivity(ctx, ComposeGreeting, greetings).Get(ctx, &allGreetings)
```

Comment:
What if a signal comes in while processing this?

Reply:
I wanted to demonstrate what happens in these signal-at-end-of-workflow cases. There's a test, Test_Signal_After_Exit(), and some code in the starter using a flag testSignalAfterExitSignal that works with this. In Go these signals are ignored, in Java the workflow goes back and processes them. 
Would it be good to add another check to len(unprocessedGreetings) == 0 before returning? I can comment why the sample does this, if you think it's worth doing so.

Reply:
> In Go these signals are ignored

This is data loss which is usually not what is wanted with workflows like this

Reply:
updated to fix this, changed the relevant test.